### PR TITLE
Windows PCA parser

### LINKED ISSUE #4558: Win 11 PCA Parser
**Description of problem:**

From the blogpost:
https://aboutdfir.com/new-windows-11-pro-22h2-evidence-of-execution-artifact/

PcaAppLaunchDic.txt contains a file path and timestamp value pair of data which provides the last time of execution of a given application.

We should parse this :)

### PR DESCRIPTION
**Pull request by Fryyyyy**
## One line description of pull request
Implements a parser for Windows AppCompat PCA (Program Compatibility Assistant)

## Description:
**Related issue (if applicable):** fixes #4558 

## Notes:
All contributions to Plaso undergo [code review](https://github.com/log2timeline/l2tdocs/blob/main/process/Code%20review%20process.asciidoc).
This makes sure that the code has appropriate test coverage and conforms to the
[Plaso style guide](https://plaso.readthedocs.io/en/latest/sources/developer/Style-guide.html).

One of the maintainers will examine your code, and may request changes. Check off the items below in
order, and then a maintainer will review your code.

## Checklist:
* [x] Automated checks (GitHub Actions, AppVeyor) pass
* [x] No new [new dependencies](https://plaso.readthedocs.io/en/latest/sources/developer/Adding-a-new-dependency.html) are required or l2tdevtools has been updated
* [x] Reviewer assigned

### CODE REVIEW

**Comment by Fryyyyy** [comment]
Replaced the test data with that generated from my own personal Win 11 machine

File: tests/parsers/text_plugins/winpca.py [hunk 1]
```diff
@@ -0,0 +1,68 @@
+#!/usr/bin/python
+# -*- coding: utf-8 -*-
+"""Tests for Win PCA log text parser plugin."""
+
+from plaso.parsers.text_plugins import winpca
+from tests.parsers.text_plugins import test_lib
+
+
+class WinPCALogTextPluginTest(test_lib.TextPluginTestCase):
+  """Tests for Apache access log text parser plugin."""
```
Comment (sydp) [inline tests/parsers/text_plugins/winpca.py:10]: Old docstring?

File: plaso/parsers/text_plugins/winpca.py [hunk 2]
```diff
@@ -0,0 +1,215 @@
...
+    _SKIP_TO_SEPARATOR.setResultsName('version') +
+    _SEPARATOR +
+    _SKIP_TO_SEPARATOR.setResultsName('program_id') +
+    _SEPARATOR +
+    pyparsing.SkipTo(pyparsing.LineEnd()).setResultsName('exit_code') +
+    _END_OF_LINE
+  )
+
+  _LINE_STRUCTURES = [
+    ('launch_dic_log_line', _LAUNCH_DIC_LOG_LINE),
```
Comment (sydp) [inline plaso/parsers/text_plugins/winpca.py:106]: Add indent

File: plaso/parsers/text_plugins/winpca.py [hunk 3]
```diff
@@ -0,0 +1,215 @@
...
+  _LAUNCH_DIC_LOG_LINE = (
+    pyparsing.SkipTo('|').setResultsName('exec') +
+    _SEPARATOR +
+    _DATE.setResultsName('date') +
+    _TIME.setResultsName('time') +
+    _END_OF_LINE
+  )
+
+  _DB0_LOG_LINE = (
+    _DATE.setResultsName('date') +
```
Comment (sydp) [inline plaso/parsers/text_plugins/winpca.py:86]: Add indent

File: plaso/parsers/text_plugins/winpca.py [hunk 4]
```diff
@@ -0,0 +1,215 @@
...
+  _TIME = pyparsing.Group(
+      _TWO_DIGITS + pyparsing.Suppress(':') +
+      _TWO_DIGITS + pyparsing.Suppress(':') +
+      _TWO_DIGITS + pyparsing.Suppress('.') +
+      _THREE_DIGITS)
+
+  _END_OF_LINE = pyparsing.Suppress(pyparsing.LineEnd())
+
+  _LAUNCH_DIC_LOG_LINE = (
+    pyparsing.SkipTo('|').setResultsName('exec') +
```
Comment (sydp) [inline plaso/parsers/text_plugins/winpca.py:78]: Add indent

File: plaso/parsers/text_plugins/winpca.py [hunk 5]
```diff
@@ -0,0 +1,215 @@
...
+  _LINE_STRUCTURES = [
+    ('launch_dic_log_line', _LAUNCH_DIC_LOG_LINE),
+    ('db0_log_line', _DB0_LOG_LINE)]
+
+  VERIFICATION_GRAMMAR = (_LAUNCH_DIC_LOG_LINE ^ _DB0_LOG_LINE)
+
+  def __init__(self):
+    """Initializes a text parser plugin."""
+    super(WinPCALogTextPlugin, self).__init__()
+    self._use_local_time = False
```
Comment (sydp) [inline plaso/parsers/text_plugins/winpca.py:114]: Does this need a docstring as an attribute?

**Review by sydp — CHANGES_REQUESTED** [review]
A couple of style nits.

File: plaso/parsers/text_plugins/winpca.py [hunk 6]
```diff
@@ -0,0 +1,215 @@
...
+
+
+class WinPCAEventData(events.EventData):
+  """Windows PCA (Program Compatibility Assistant) event data.
+
+  Attributes:
+    body (str): message body.
+    description (str): Description of the executable.
+    exit_code (str): Final result of the execution.
+    last_written_time (dfdatetime.DateTimeValues): entry last written date and
```
Comment (sydp) [inline plaso/parsers/text_plugins/winpca.py:21]: if i understand the blog post correctly, should this be ```execution_time``` or something similar?

File: plaso/parsers/text_plugins/winpca.py [hunk 7]
```diff
@@ -0,0 +1,215 @@
...
+    self.vendor = None
+    self.version = None
+
+class WinPCALogTextPlugin(interface.TextPlugin):
+  """Text parser plugin for Windows PCA Log files."""
+
+  NAME = 'winpca'
+  DATA_FORMAT = 'Windows PCA log file'
+
+  _SEPARATOR = pyparsing.Suppress('|')
```
Comment (sydp) [inline plaso/parsers/text_plugins/winpca.py:49]: Maybe _DELIMITER?

File: data/presets.yaml [hunk 8]
```diff
@@ -165,6 +165,7 @@ parsers:
 - olecf/olecf_automatic_destinations
 - recycle_bin
 - winevtx
+- text/winpca
```
Comment (joachimmetz) [inline data/presets.yaml:168]: please sort alphabetically

File: plaso/parsers/text_plugins/winpca.py [hunk 9]
```diff
@@ -0,0 +1,220 @@
+# -*- coding: utf-8 -*-
+"""Text parser plugin for Windows PCA Log files."""
```
Comment (joachimmetz) [inline plaso/parsers/text_plugins/winpca.py:2]: mention what PCA stands for

File: plaso/parsers/text_plugins/winpca.py [hunk 10]
```diff
@@ -0,0 +1,220 @@
...
+  DATA_TYPE = 'windows:pca_log:entry'
+
+  def __init__(self):
+    """Initializes event data."""
+    super(WinPCAEventData, self).__init__(data_type=self.DATA_TYPE)
+    self.body = None
+    self.description = None
+    self.exit_code = None
+    self.last_execution_time = None
+    self.program_id = None
```
Comment (joachimmetz) [inline plaso/parsers/text_plugins/winpca.py:38]: style nit: don't use abbreviations

File: plaso/parsers/text_plugins/winpca.py [hunk 11]
```diff
@@ -0,0 +1,220 @@
...
+
+
+class WinPCAEventData(events.EventData):
+  """Windows PCA (Program Compatibility Assistant) event data.
+
+  Attributes:
+    body (str): message body.
+    description (str): Description of the executable.
+    exit_code (str): Final result of the execution.
+    last_execution_time (dfdatetime.DateTimeValues): entry last written date and
```
Comment (joachimmetz) [inline plaso/parsers/text_plugins/winpca.py:21]: last written or execution time?

File: plaso/parsers/text_plugins/winpca.py [hunk 12]
```diff
@@ -0,0 +1,220 @@
...
+        time.
+    program_id (str): Program ID.
+    run_status (str): Execution status.
+    vendor (str): Software vendor.
```
Comment (joachimmetz) [inline plaso/parsers/text_plugins/winpca.py:25]: vendor of what ? the program that ran?

File: plaso/parsers/text_plugins/winpca.py [hunk 13]
```diff
@@ -0,0 +1,220 @@
...
+    version (str): Software version.
```
Comment (joachimmetz) [inline plaso/parsers/text_plugins/winpca.py:26]: version of what ? the program that ran?

**Comment by joachimmetz** [comment]
@Fryyyyy a design decision question: why a pyparsing parser and not a DSVParser parser ?

File: tests/parsers/text_plugins/winpca.py [hunk 14]
```diff
@@ -0,0 +1,68 @@
...
+            '.702.1670.9453_x64__8wekyb3d8bbwe\\msteamsupdate.exe'),
+        'description': 'microsoft teams updater',
+        'vendor': 'microsoft corporation',
+        'version': '22287.702.1670.9453',
+        'program_id': '0006132687b1e64961e910ee21d4352afe0800000904',
+        'exit_code': 'Abnormal process exit with code 0x4c7',
+        'last_execution_time': '2022-11-14T23:37:11.789+00:00'}
+
+    event_data = storage_writer.GetAttributeContainerByIndex('event_data', 0)
+    self.CheckEventData(event_data, expected_event_values)
```
Comment (joachimmetz) [inline tests/parsers/text_plugins/winpca.py:68]: missing `__main__`

File: plaso/parsers/text_plugins/winpca.py [hunk 15]
```diff
@@ -0,0 +1,220 @@
...
+    Args:
+      parser_mediator (ParserMediator): mediates interactions between parsers
+          and other components, such as storage and dfVFS.
+      text_reader (EncodedTextReader): text reader.
+
+    Returns:
+      bool: True if this is the correct parser, False otherwise.
+    """
+    try:
+      self._VerifyString(text_reader.lines)
```
Comment (joachimmetz) [inline plaso/parsers/text_plugins/winpca.py:211]: how costly is this verification ?

File: plaso/parsers/text_plugins/winpca.py [hunk 16]
```diff
@@ -0,0 +1,220 @@
...
+    super(WinPCAEventData, self).__init__(data_type=self.DATA_TYPE)
+    self.body = None
+    self.description = None
+    self.exit_code = None
+    self.last_execution_time = None
+    self.program_id = None
+    self.run_status = None
+    self.vendor = None
+    self.version = None
+
```
Comment (joachimmetz) [inline plaso/parsers/text_plugins/winpca.py:42]: add empty line

**Comment by Fryyyyy** [comment]
Ported to DSV and made sure it doesn't parse a Bodyfile :)

File: plaso/parsers/text_plugins/winpca.py [hunk 17]
```diff
@@ -0,0 +1,192 @@
...
+  DELIMITER = '|'
+
+  COLUMNS = ()
+  _MINIMUM_NUMBER_OF_COLUMNS = None
+
+  def _ParseDateTime(self, date_time_string):
+    """Parses date and time elements of a log line.
+
+    Args:
+      structure (pyparsing.ParseResults): tokens from a parsed log line.
```
Comment (joachimmetz) [inline plaso/parsers/text_plugins/winpca.py:58]: docstring doesn't match actual argument

File: plaso/parsers/text_plugins/winpca.py [hunk 18]
```diff
@@ -0,0 +1,188 @@
...
+      row (dict[str, str]): fields of a single row, as specified in COLUMNS.
+
+    Returns:
+      bool: True if this is the correct parser, False otherwise.
+    """
+    if len(row) < self._MINIMUM_NUMBER_OF_COLUMNS:
+      return False
+
+    try:
+      _ = self._ParseDateTime(row['datetime'])
```
Comment (joachimmetz) [inline plaso/parsers/text_plugins/winpca.py:112]: You can safely remove `_ =`

File: plaso/parsers/text_plugins/winpca.py [hunk 19]
```diff
@@ -0,0 +1,188 @@
...
+  COLUMNS = ['program', 'datetime']
+
+  _MINIMUM_NUMBER_OF_COLUMNS = 2
+
+  def ParseRow(self, parser_mediator, row_offset, row):
+    """Parses a line of the log file and produces events.
+
+    Args:
+      parser_mediator (ParserMediator): mediates interactions between parsers
+          and other components, such as storage and dfvfs.
```
Comment (joachimmetz) [inline plaso/parsers/text_plugins/winpca.py:134]: dfvfs => dfVFS

File: plaso/parsers/text_plugins/winpca.py [hunk 20]
```diff
@@ -0,0 +1,188 @@
...
+    try:
+      _ = self._ParseDateTime(row['datetime'])
+    except (AttributeError, ValueError):
+      return False
+
+    return '\\' in row['program']
+
+
+class WinPCADicParser(WinPCABaseParser):
+  """Parses the Windows Program Compatibility Assistant DIC files"""
```
Comment (joachimmetz) [inline plaso/parsers/text_plugins/winpca.py:120]: add dot at end of sentence

File: plaso/parsers/text_plugins/winpca.py [hunk 21]
```diff
@@ -0,0 +1,188 @@
...
+    'version', 'program_id', 'exit_code']
+
+  _MINIMUM_NUMBER_OF_COLUMNS = 8
+
+  def ParseRow(self, parser_mediator, row_offset, row):
+    """Parses a line of the log file and produces events.
+
+    Args:
+      parser_mediator (ParserMediator): mediates interactions between parsers
+          and other components, such as storage and dfvfs.
```
Comment (joachimmetz) [inline plaso/parsers/text_plugins/winpca.py:166]: dfvfs => dfVFS

File: plaso/parsers/text_plugins/winpca.py [hunk 22]
```diff
@@ -0,0 +1,188 @@
...
+    date_time.is_local_time = False
+
+    return date_time
+
+  def VerifyRow(self, parser_mediator, row):
+    """Verifies if a line of the file is in the expected format.
+
+    Args:
+      parser_mediator (ParserMediator): mediates interactions between parsers
+          and other components, such as storage and dfvfs.
```
Comment (joachimmetz) [inline plaso/parsers/text_plugins/winpca.py:102]: dfvfs => dfVFS

File: plaso/parsers/text_plugins/winpca.py [hunk 23]
```diff
@@ -0,0 +1,188 @@
...
+
+    event_data = WinPCAEventData()
+    event_data.body = row['program']
+    event_data.last_execution_time = date_time
+
+    parser_mediator.ProduceEventData(event_data)
+
+
+class WinPCADB0Parser(WinPCABaseParser):
+  """Parses the Windows Program Compatibility Assistant DB0 files"""
```
Comment (joachimmetz) [inline plaso/parsers/text_plugins/winpca.py:150]: add dot at end of sentence

**Review by joachimmetz — CHANGES_REQUESTED** [review]
A couple of small nits remaining.

File: plaso/parsers/text_plugins/winpca.py [hunk 24]
```diff
@@ -0,0 +1,188 @@
...
+    self.description = None
+    self.exit_code = None
+    self.last_execution_time = None
+    self.program_identifier = None
+    self.run_status = None
+    self.vendor = None
+    self.version = None
+
+
+# TODO: Check for conflicts with Bodyfile.
```
Comment (joachimmetz) [inline plaso/parsers/text_plugins/winpca.py:42]: I assume this is a TODO? And assume this needs to be resolved before the code can be merged?

File: plaso/parsers/text_plugins/winpca.py [hunk 25]
```diff
@@ -0,0 +1,188 @@
...
+      row_offset (int): offset of the line from which the row was extracted.
+      row (dict[str, str]): fields of a single row, as specified in COLUMNS.
+    """
+    date_time = self._ParseDateTime(row['datetime'])
+    if date_time is None:
+      return
+
+    event_data = WinPCAEventData()
+    event_data.body = row['program']
+    event_data.last_execution_time = date_time
```
Comment (joachimmetz) [inline plaso/parsers/text_plugins/winpca.py:144]: Change to `= self._ParseDateTime(row['datetime'])` and remove block above. Change `place_holder_event:` to false in the timeliner configuration if WinPCAEventData without a last_execution_time should not yield events.

File: plaso/parsers/text_plugins/winpca.py [hunk 26]
```diff
@@ -0,0 +1,188 @@
...
+
+
+class WinPCADB0Parser(WinPCABaseParser):
+  """Parses the Windows Program Compatibility Assistant DB0 files"""
+
+  NAME = 'winpca_db0'
+  DATA_FORMAT = 'Windows PCA log file'
+
+  COLUMNS = [
+    'datetime', 'run_status', 'program', 'description', 'vendor',
```
Comment (joachimmetz) [inline plaso/parsers/text_plugins/winpca.py:156]: Please use 4 space continuation indentation.

File: plaso/parsers/text_plugins/winpca.py [hunk 27]
```diff
@@ -0,0 +1,188 @@
...
+  def ParseRow(self, parser_mediator, row_offset, row):
+    """Parses a line of the log file and produces events.
+
+    Args:
+      parser_mediator (ParserMediator): mediates interactions between parsers
+          and other components, such as storage and dfvfs.
+      row_offset (int): offset of the line from which the row was extracted.
+      row (dict[str, str]): fields of a single row, as specified in COLUMNS.
+    """
+    date_time = self._ParseDateTime(row['datetime'])
```
Comment (joachimmetz) [inline plaso/parsers/text_plugins/winpca.py:170]: similar remark as before

File: plaso/parsers/text_plugins/winpca.py (hunk 24 above)
Comment (Fryyyyy) [reply plaso/parsers/text_plugins/winpca.py:42]: Done

**Comment by joachimmetz** [comment]
A couple of things for feedback, I'll make the changes

* `body` is not very descriptive name, and `program` seems to be more appropriate
* seeing this uses the DSV parser it is technically not a text plugin (as in using the text plugin base class)

File: plaso/parsers/winpca.py [hunk 28]
```diff
@@ -0,0 +1,176 @@
...
+
+    Returns:
+      bool: True if this is the correct parser, False otherwise.
+    """
+    if len(row) < self._MINIMUM_NUMBER_OF_COLUMNS:
+      return False
+
+    try:
+      self._ParseDateTime(row['datetime'])
+    except (AttributeError, ValueError):
```
Comment (joachimmetz) [inline plaso/parsers/winpca.py:109]: this should be ParseError.

File: plaso/parsers/winpca.py [hunk 29]
```diff
@@ -0,0 +1,176 @@
...
+      date_time_string (string): Date time values extracted from the logfile.
+
+    Returns:
+      dfdatetime.TimeElements: date and time value.
+
+    Raises:
+      ParseError: if a valid date and time value cannot be derived from
+          the time elements.
+    """
+    if (date_time_string[4] != '-' or date_time_string[7] != '-' or
```
Comment (joachimmetz) [inline plaso/parsers/winpca.py:66]: This can be replaced by CopyFromDateTimeString

File: plaso/parsers/winpca.py [hunk 30]
```diff
@@ -0,0 +1,176 @@
...
+      parser_mediator (ParserMediator): mediates interactions between parsers
+          and other components, such as storage and dfVFS.
+      row_offset (int): offset of the line from which the row was extracted.
+      row (dict[str, str]): fields of a single row, as specified in COLUMNS.
+    """
+    event_data = WindowsPCAEventData()
+    event_data.description = row['description']
+    event_data.executable = row['program']
+    event_data.exit_code = row['exit_code']
+    event_data.last_execution_time = self._ParseDateTime(row['datetime'])
```
Comment (joachimmetz) [inline plaso/parsers/winpca.py:140]: not catching the exception will cause the parser to stop parsing

**Review by joachimmetz — APPROVED** [review]
LGTM

**Pull request merged**